# DATA PRE-PROCESSING

## DESCRIPTION
Before applying provided R scripts, Bulk RNA-seq and DNase-seq data were pre-processed using custom pipelines [**Workflow_RNA-seq**](https://github.com/JosephLeger/Workflow_RNA-seq) ([v1.0.0](https://github.com/JosephLeger/Workflow_RNA-seq/releases/tag/v1.0.0)) and [**Workflow_ChIP-like**](https://github.com/JosephLeger/Workflow_ChIP-like) ([v1.0.0](https://github.com/JosephLeger/Workflow_ChIP-like/releases)). Single cell RNA-seq data were pre-processed using **CellRanger**.
To reproduce the same pre-processing steps, we provided here the exact command lines used for these respective custom pipelines, or the equivalent command line syntax while using directly tools.  
Mapping and annotation were realized using Ensembl [GRCm39](https://ftp.ensembl.org/pub/release-110/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa.gz) and [GTF release 108](https://ftp.ensembl.org/pub/release-108/gtf/mus_musculus/Mus_musculus.GRCm39.108.gtf.gz).

## SINGLE CELL RNA-SEQ
### Requirements
```
Name                        Version
cellranger                  7.0.1
```

### Using directly tools
```bash
# Filter GTF file from Ensembl
cellranger mkgtf ../Genome/Mus_musculus.GRCm39.108.gtf Mus_musculus.GRCm39.108.filtered.gtf \
--attribute=gene_biotype:protein_coding \
--attribute=gene_biotype:lncRNA \
--attribute=gene_biotype:antisense \
--attribute=gene_biotype:IG_LV_gene \
--attribute=gene_biotype:IG_V_gene \
--attribute=gene_biotype:IG_V_pseudogene \
--attribute=gene_biotype:IG_D_gene \
--attribute=gene_biotype:IG_J_gene \
--attribute=gene_biotype:IG_J_pseudogene \
--attribute=gene_biotype:IG_C_gene \
--attribute=gene_biotype:IG_C_pseudogene \
--attribute=gene_biotype:TR_V_gene \
--attribute=gene_biotype:TR_V_pseudogene \
--attribute=gene_biotype:TR_D_gene \
--attribute=gene_biotype:TR_J_gene \
--attribute=gene_biotype:TR_J_pseudogene \
--attribute=gene_biotype:TR_C_gene

# Prepare cellranger reference genome
cellranger mkref --genome=mm39 --fasta=../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa --genes=Mus_musculus.GRCm39.108.filtered.gtf

# Count reads
cellranger count --id ${sample} --transcriptome ../Ref/refdata-cellranger-mm39/mm39 --fastqs Raw/${sample} --sample ${sample} --jobmode sge --maxjobs 10 --disable-ui
```
Note : the GTF file generated by ```mkgtf``` is also used for peaks annotation.  


## DNASE-SEQ
### Requirements
```
Name                        Version
fastqc                      0.11.9
multiqc                     1.13
trimmomatic                 0.39
bbmap                       39.00
bowtie2                     2.5.1
samtools                    1.15.1
picard                      2.23.5
bedtools                    2.30.0
ucsc-bedgraphtobigwig       377
gcc                         11.2.0
macs2                       2.2.7.1
homer                       4.11
```

### Using [Custom Pipeline](https://github.com/JosephLeger/Workflow_ChIP-like)
```bash
# Quality Check of raw FASTQ files
sh 1_QC.sh Raw

# Trimming to remove adapters and duplicated reads
sh 2_Trim.sh -U 'Both' -S 4:15 -L 5 -T 5 -M 36 -I ../Ref/TruSeq3-SE_NexteraPE-PE.fa:2:30:7 -D True SE Raw

# Quality Check after trimming
sh 1_QC.sh Trimmed/Trimmomatic

# Mapping using Bowtie2
sh 3_Bowtie2.sh SE Trimmed/Trimmomatic ../Ref/refdata-Bowtie2-mm39/mm39

# Filtering
sh 4_BowtieCheck.sh -N _sorted -T 10 -R false Mapped/mm39/BAM 

# Merge BAM files and BW generation
sh mergeBAM.sh -N _Clum_Trimmed_sorted_filtered -R true Mapped/mm39/BAM SRA_list_DNAse-seq.csv
sh BAM2BW.sh -N _merged -F bigwig -R true Mapped/mm39/BAM

# Peak calling with MACS2
sh 5_PeakyFinders.sh -U 'MACS2' -N _merged Mapped/mm39/BAM
sh 5_PeakyFinders.sh -U 'MACS2' -N _Clum_Trimmed_sorted_filtered Mapped/mm39/BAM

# Motif enrichment analysis and peak annotation with HOMER from patterns identified by R DiffBind using filtered GTF from scRNA-seq analysis
sh 6_Annotate.sh -N '' -R 200 -L '8,10,12' -A true -M true NFIL3_dev_ILC/Peaks ../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa ../Ref/refdata-cellranger-mm39/Mus_musculus.GRCm39.108.filtered.gtf
```

### Using directly tools
```bash
# Trimming to remove duplicated reads and adapters
clumpify.sh in=${file}.fastq out=${file}_Clum.fastq dedupe=t subs=0
trimmomatic SE -threads 4 ${file}_Clum.fastq ${file}_Clum_Trimmed.fastq SLIDINGWINDOW:4:15 LEADING:5 TRAILING:5 MINLEN:36 ILLUMINACLIP:../Ref/TruSeq3-SE_NexteraPE-PE.fa:2:30:7

# Mapping using Bowtie2
bowtie2 -p 2 -N 0 -x ../Ref/refdata-Bowtie2-mm39/mm39 -U ${file}.fastq -S Mapped/mm39/SAM/${file}.sam
picard SortSam INPUT=Mapped/mm39/SAM/${file}.sam OUTPUT=Mapped/mm39/BAM/${file}_sorted.bam \
VALIDATION_STRINGENCY=LENIENT TMP_DIR=tmp SORT_ORDER=coordinate

# Filtering
samtools view -h ${file}.bam | samtools view -b -Sq 10  > ${file}_filtered.bam
samtools index ${file}_filtered.bam ${bam}_filtered.bai

# Merge BAM files and BW generation
samtools merge -o Mapped/mm39/BAM/${file}.bam ${files_to_merge}
samtools index Mapped/mm39/BAM/${file}.bam Mapped/mm39/BAM/${file}.bai
bamCoverage --outFileFormat bigwig -b ${file}.bam -o Mapped/mm39/BIGWIG/${file}.bw

# Peak calling with MACS2
macs2 callpeak -t ${file}.bam -f BAM -g 1.87e9 --nomodel --shift 50 --extsize 100 -n ${file}_summits.bed --outdir MACS2/Peaks

# Motif enrichment analysis and peak annotation with HOMER from patterns identified by R DiffBind using filtered GTF from scRNA-seq analysis
annotatePeaks.pl ${file}.bed ../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa -gtf ../Ref/refdata-cellranger-mm39/Mus_musculus.GRCm39.108.filtered.gtf > ${file}_annotated.txt
findMotifsGenome.pl ${file}.bed ../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa ${outdir} -size 200 -len '8,10,12' -S 40
```


## BULK RNA-SEQ
### Requirements
```
Name                        Version
fastqc                      0.11.9
multiqc                     1.13
trimmomatic                 0.39
rsem                        1.3.2
star                        2.7.5a
```

### Using [Custom Pipeline](https://github.com/JosephLeger/Workflow_RNA-seq)
```bash
# Quality Check of raw FASTQ files
sh QC.sh Raw

# Trimming to remove adapters
# For Batch 1 from GSE291075 and Batch 3 from GSE291076
sh Trim.sh -S 4:15 -L 3 -T 3 -M 36 -I ../Ref/NexteraPE-PE_Clontech-TTT.fa:2:30:10 PE Raw
# For Batch 2 from GSE291075
sh Trim.sh -S 4:15 -L 5 -T 3 -M 36 -I ../Ref/NexteraPE-PE_Clontech-TTT.fa:2:30:10 PE Raw
# For Batch 4 from GSE291316, no Trimming step was necessary

# Quality Check after trimming
sh QC.sh Trimmed/Trimmomatic/Paired

# RSEM alignment
sh RSEM.sh PE Trimmed/Trimmomatic/Paired ../Ref/refdata-RSEM-mm39.108/mm39.108
```

### Using directly tools
```bash
# Trimming to remove adapters
# For Batch 1 from GSE291075 and Batch 3 from GSE291076
trimmomatic PE -threads 4 ${R1} ${R2} \
Trimmed/Trimmomatic/Paired/${P1} Trimmed/Trimmomatic/Unpaired/${U1} \
Trimmed/Trimmomatic/Paired/${P2} Trimmed/Trimmomatic/Unpaired/${U2} \
ILLUMINACLIP:../Ref/NexteraPE-PE_Clontech-TTT.fa:2:30:10 SLIDINGWINDOW:4:15 LEADING:3 TRAILING:3 MINLEN:36 
# For Batch 2 from GSE291075
trimmomatic PE -threads 4 ${R1} ${R2} \
Trimmed/Trimmomatic/Paired/${P1} Trimmed/Trimmomatic/Unpaired/${U1} \
Trimmed/Trimmomatic/Paired/${P2} Trimmed/Trimmomatic/Unpaired/${U2} \
ILLUMINACLIP:../Ref/NexteraPE-PE_Clontech-TTT.fa:2:30:10 SLIDINGWINDOW:4:15 LEADING:5 TRAILING:3 MINLEN:36
# For Batch 4 from GSE291316, no Trimming step was necessary

# RSEM alignment for Batch 1-3 from GSE291075 and GSE291076
rsem-calculate-expression -p 8 --paired-end --star --star-gzipped-read-file ${R1} ${R2} ../Ref/refdata-RSEM-mm39.108/mm39.108 ${output}
# RSEM alignment for Batch 4 from GSE291316
rsem-calculate-expression -p 8 --star --star-gzipped-read-file ${file} ../Ref/refdata-RSEM-mm39.108/mm39.108 ${output}
```


## CUT&RUN
### Requirements
```
Name                        Version
fastqc                      0.11.9
multiqc                     1.13
trimmomatic                 0.39
bbmap                       39.00
bowtie2                     2.5.1
samtools                    1.15.1
picard                      2.23.5
bedtools                    2.30.0
ucsc-bedgraphtobigwig       377
gcc                         11.2.0
homer                       4.11
```

### Using [Custom Pipeline](https://github.com/JosephLeger/Workflow_ChIP-like)
```bash
# Quality Check of raw FASTQ files
sh 1_QC.sh Raw

# Trimming to remove adapters and duplicated reads
sh 2_Trim.sh -U 'Both' -S 4:15 -L 3 -T 3 -M 0 -I ../Ref/BGI_Adapters.fa:2:30:10 -D True SE Raw

# Quality Check after trimming
sh 1_QC.sh Trimmed/Trimmomatic

# Mapping using Bowtie2
sh 3_Bowtie2.sh -M henikoff SE Trimmed/Trimmomatic ../Ref/refdata-Bowtie2-mm39/mm39

# Filtering
sh 4_BowtieCheck.sh -N _sorted -T 10 -R false Mapped/mm39/BAM 

# Merge BAM files and BW generation
sh mergeBAM.sh -N _Clum_Trimmed_sorted_filtered -R true Mapped/mm39/BAM SRA_list_CUTnRUN.csv
sh BAM2BW.sh -N _merged -F bigwig -M RPKM -R true Mapped/mm39/BAM

# Peak calling with HOMER
sh 5_PeakyFinders.sh -U 'HOMER' -M factor -S 50 -N _merged Mapped/mm39/BAM

# Motif enrichment analysis and peak annotation with HOMER from patterns identified by R DiffBind using filtered GTF from scRNA-seq analysis
sh 6_Annotate.sh -N '' -R 200 -L '8,10,12' -A true -M true NFIL3_dev_ILC/Peaks ../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa ../Ref/refdata-cellranger-mm39/Mus_musculus.GRCm39.108.filtered.gtf
```

### Using directly tools
```bash
# Trimming to remove duplicated reads and adapters
clumpify.sh in=${file}.fastq out=${file}_Clum.fastq dedupe=t subs=0
trimmomatic SE -threads 4 ${file}_Clum.fastq ${output}_Clum_Trimmed.fastq SLIDINGWINDOW:4:15 LEADING:5 TRAILING:5 MINLEN:36 ../Ref/TruSeq3-SE_NexteraPE-PE.fa:2:30:7

# Mapping using Bowtie2
bowtie2-p 2 --end-to-end --very-sensitive --no-mixed --no-discordant -x ../Ref/refdata-Bowtie2-mm39/mm39 -U ${file} -S Mapped/mm39/SAM/${file}.sam
picard SortSam INPUT=Mapped/mm39/SAM/${file}.sam OUTPUT=Mapped/mm39/BAM/${file}_sorted.bam \
VALIDATION_STRINGENCY=LENIENT TMP_DIR=tmp SORT_ORDER=coordinate

# Filtering
samtools view -h ${file}.bam | samtools view -b -Sq 10  > ${bam}_filtered.bam
samtools index ${bam}_filtered.bam ${bam}_filtered.bai

# Peak calling with HOMER
makeTagDirectory HOMER/Tags/${file}/mm39/BAM/${file}.bam -fragLength 50 -single 
makeUCSCfile HOMER/Tags/${file} -o auto
findPeaks HOMER/Tags/${file}.bam -style factor -o HOMER/Peaks/${file}/${file}_peaks.txt -L 4 -C 2 -tagThreshold 2 -i HOMER/Tags/${input_file}.bam -F 4
grep -v '^#' HOMER/Peaks/${file}/${file}_peaks.txt | awk -v OFS='\t' '{print $2,$3,$4,$1,$8,$5}' | bedtools sort > HOMER/Peaks/${file}/${file}_peaks.bed 

# Motif enrichment analysis and peak annotation with HOMER from patterns identified by R DiffBind using filtered GTF from scRNA-seq analysis
annotatePeaks.pl ${file}.bed ../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa -gtf ../Ref/refdata-cellranger-mm39/Mus_musculus.GRCm39.108.filtered.gtf > ${file}_annotated.txt
findMotifsGenome.pl ${file}.bed ../Ref/Mus_musculus.GRCm39.dna_sm.primary_assembly.fa ${outdir} -size 200 -len '8,10,12' -S 40
```




